docs/docs/dev_docs/contracts/syntax/functions.md

```diff
@@ -2,53 +2,234 @@
+
+### Data Visibility
+Data visibility is used to describe whether the data (or state) used in a function is generally accessible (public) or on a need to know basis (private). Functions with public data visibility is executed by the sequencer, and functions with private data visibility are executed by the user. For more information on why this is the case, see [communication](../../../concepts/foundation/communication/public_private_calls.md).
+
+In the coming sections, we are going to see how these two "types" co-exists and interact.
+
+### Function visibility
+This is the kind of visibility you are more used to seeing in Solidity and more "normal" programming languages. It is used to describe whether a function is callable from other contracts, or only from within the same contract.
+
+By default, all functions are callable from other contracts, similarly to the Solidity `public` visibility. To make them only callable from the contract itself, you can mark them as `internal`.
+
+A good place to use `internal` is when you want a private function to be able to alter public state. While as mentioned above, private functions cannot do this directly, they are able to call public functions, and by making these internal, we can ensure that this state manipulating function is only callable from our private function.
+
+:::danger
+Note that non-internal functions could be used directly as an entry-point, which currently means that the `msg_sender` would be `0`, so for now, using address `0` as a burn address is not recommended.
+:::
+
 ## `constructor`
 
 - A special `constructor` function MUST be declared within a contract's scope.
 - A constructor doesn't have a name, because its purpose is clear: to initialise contract state.
 - In Aztec terminology, a constructor is always a '`private` function' (i.e. it cannot be a `public` function, in the current version of the sandbox it cannot call public functions either).
 - A constructor behaves almost identically to any other function. It's just important for Aztec to be able to identify this function as special: it may only be called once, and will not be deployed as part of the contract.
 
-An example of a constructor is as follows:
+An example of a somewhat boring constructor is as follows:
+
+#include_code empty-constructor /yarn-project/noir-contracts/src/contracts/public_token_contract/src/main.nr rust
+
+Although you can have a constructor that does nothing, you might want to do something with it, such as setting the deployer as an owner.
+
 #include_code constructor /yarn-project/noir-contracts/src/contracts/escrow_contract/src/main.nr rust
 
-In this example (taken from an escrow contract), the constructor sets the deployer as an `owner`.
+:::danger
+It is not possible to call public functions from within a constructor. Beware that the compiler will not throw an error if you do, but the execution will fail!
```

Comment:
```suggestion
It is not possible, at this time, to call public functions from within a constructor. Beware that the compiler will not throw an error if you do, but the execution will fail!
```
1. add a note on why
2. say that this is just temporary and working on a fix?

Reply:
It is already mentioned in the small list under the constructor, I just created the danger box for extra eye catching.

Reply:
Or, the why is not mentioned in there. But where thinking showing that into the limitations page to not bloat in here?

Reply:
To not be redundant here, I suggest adding it to the Limitations page and adding a link here.